### ISSUE TITLE
Sample Code for Cutting/Shorten Videos

### ISSUE THREAD

**Issue by iAviatorJose** [sannies/mp4parser#51]
I'm new to this lib, and i want to cut videos and encode and decode a video from .MOV (ios) to .mp4
Can you post a sample code demonstrating the cutting of videos.

**Comment by iAviatorJose** [comment]
well i did get the sample to crop the video by following your SimpleShortenExample but i see the when i set the start and end time of the frame it does get cropped exactly from that time.

**Comment by sannies** [comment]
Could you rephrase? I don't really understand the probem.

2015-02-25 13:05 GMT+02:00 Jose Varghese <notifications@github.com>:

**Comment by iAviatorJose** [comment]
public void Mp4CutProcessBegin(){
        try{
            progressBar.setVisibility(View.VISIBLE);
            File sdCard = Environment.getExternalStorageDirectory();
            File file1 = new File(tv_video_url.getText().toString());
            Movie movie = MovieCreator.build(file1.getPath());

            List<Track> tracks = movie.getTracks();
            movie.setTracks(new LinkedList<Track>());
            // remove all tracks we will create new tracks from the old

            double startTime1 = Double.parseDouble(et_start_time.getText().toString());  // What Time Should i set here and in which unit of time.
            double endTime1 = Double.parseDouble(et_end_time.getText().toString());   // What Time Should i set here and in which unit of time.

            boolean timeCorrected = false;

            // Here we try to find a track that has sync samples. Since we can only start decoding
            // at such a sample we SHOULD make sure that the start of the new fragment is exactly
            // such a frame
            for (Track track : tracks) {
                if (track.getSyncSamples() != null && track.getSyncSamples().length > 0) {
                    if (timeCorrected) {
                        // This exception here could be a false positive in case we have multiple tracks
                        // with sync samples at exactly the same positions. E.g. a single movie containing
                        // multiple qualities of the same video (Microsoft Smooth Streaming file)

                        throw new RuntimeException("The startTime has already been corrected by another track with SyncSample. Not Supported.");
                    }
                    startTime1 = correctTimeToSyncSample(track, startTime1, false);
                    endTime1 = correctTimeToSyncSample(track, endTime1, true);
                    timeCorrected = true;
                }
            }

            for (Track track : tracks) {
                long currentSample = 0;
                double currentTime = 0;
                double lastTime = -1;
                long startSample1 = -1;
                long endSample1 = -1;

                for (int i = 0; i < track.getSampleDurations().length; i++) {
                    long delta = track.getSampleDurations()[i];


                    if (currentTime > lastTime && currentTime <= startTime1) {
                        // current sample is still before the new starttime
                        startSample1 = currentSample;
                    }
                    if (currentTime > lastTime && currentTime <= endTime1) {
                        // current sample is after the new start time and still before the new endtime
                        endSample1 = currentSample;
                    }
                    lastTime = currentTime;
                    currentTime += (double) delta / (double) track.getTrackMetaData().getTimescale();
                    currentSample++;
                }
                movie.addTrack(new AppendTrack(new CroppedTrack(track, startSample1, endSample1)));
            }
            long start1 = System.currentTimeMillis();
            Container out = new DefaultMp4Builder().build(movie);
            long start2 = System.currentTimeMillis();
            File output = new File(sdCard+"/output.mp4");
            if(output.exists()){
                output.delete();
            }
            FileOutputStream fos = new FileOutputStream(sdCard+"/output.mp4"); //String.format("output-%f-%f.mp4", startTime1, endTime1)
            FileChannel fc = fos.getChannel();
            out.writeContainer(fc);

            fc.close();
            fos.close();
            long start3 = System.currentTimeMillis();
            System.err.println("Building IsoFile took : " + (start2 - start1) + "ms");
            System.err.println("Writing IsoFile took  : " + (start3 - start2) + "ms");

        } catch (Exception e){
            e.printStackTrace();
        }

        progressBar.setVisibility(View.GONE);

    }

**Comment by iAviatorJose** [comment]
When i specify the start and end time for cropping of the video the resultant video does not get cropped from the specified start and end time.
Ex. if the video is of 1 min and 
say, start_time = 4sec and end_time = 20 sec 
Then the resultant video doesn't get cropped from 4 sec

**Comment by sannies** [comment]
A video can only start playing from an I-frame and if you stop playing
right before an I-frame you are also on the safe side. Do you have an
I-frame at the start time?

2015-02-25 13:14 GMT+02:00 Jose Varghese <notifications@github.com>:

**Comment by iAviatorJose** [comment]
Sir, I dont know how to figure that out whether a frame is I,b or P frame but 
I have one more question in what Unit of time Should Start_Time and End_Time Should be.
I.E. whether it should be in min/Seconds/Milliseconds/Microseconds

**Comment by sunshinetpu** [comment]
So finally is there a way to cut video exactly where we want? Even at the middle of an Iframe? Because the length of an Iframe is 10 seconds, so if video can only start playing from an Iframe, we would get a big error.

**Comment by iAviatorJose** [comment]
@sunshinetpu no dear @sannies Sebastian please give your valuable inputs.

**Comment by emce** [comment]
I have a similiar error: I have a movie with 5.6 second length, and if I try to crop it to 1 sec, I'm always gettting 2.4sec trimmed movie length. And if I set start trim point on value lower than 2.4sec, it is set on 0.0sec. I use code from shorten example...

**Comment by emce** [comment]
Basing on your code from here https://github.com/sannies/mp4parser/issues/57 I got:
```
06-15 14:05:45.582  24549-24615/com.shotclip.android E/tracks size﹕ track size =2
06-15 14:05:45.582  24549-24615/com.shotclip.android E/﹕ [ 06-15 14:05:45.582 24549:24615 E/         ]
06-15 14:05:45.582  24549-24615/com.shotclip.android E/﹕ [ 06-15 14:05:45.582 24549:24615 E/         ]
.
.
. 
06-15 14:05:45.584  24549-24615/com.shotclip.android E/﹕ [ 06-15 14:05:45.584 24549:24615 E/         ]
06-15 14:05:45.584  24549-24615/com.shotclip.android W/System.err﹕ corrected starttime from 8.0 to 9.599999999999971
06-15 14:05:45.584  24549-24615/com.shotclip.android E/﹕ [ 06-15 14:05:45.584 24549:24615 E/         ]
.
.
.
06-15 14:05:45.587  24549-24615/com.shotclip.android E/﹕ [ 06-15 14:05:45.587 24549:24615 E/         ]
06-15 14:05:45.587  24549-24615/com.shotclip.android W/System.err﹕ corrected endTime from 10.01 to 9.599999999999971
06-15 14:05:45.587  24549-24615/com.shotclip.android E/track list﹕ start time=9.599999999999971 end Time =9.599999999999971
06-15 14:05:45.587  24549-24615/com.shotclip.android E/TRACKLIST﹕ size track =250
06-15 14:05:45.588  24549-24615/com.shotclip.android E/TRACKLIST﹕ size track =431
06-15 14:05:45.588  24549-24615/com.shotclip.android E/TRACKLIST2﹕ startSample =413 endSample=413
```

**Comment by sannies** [comment]
You can only cut video at I-frame borders. As samples are dependant on each
other you cannot just start the video at any frame.  This is why cutting is
not precise.

Michał Ćwikliński <notifications@github.com> schrieb am Mo., 15. Juni 2015
14:12:

**Comment by chad187** [comment]
How do you find an Iframe and how long is an Iframe?  For me private static double correctTimeToSyncSample(Track track, double cutHere, boolean next)  is making my start and end time the same.  Should it work that way?

**Comment by sannies** [comment]
Hi,

well with the information given I can only guess. But it seems that start
and end are so close together (or the I-frame so far apart) that for each
one the I-frame in closest proximity is the same I-frame. Try to increase
the distance between start and end and see how it works out.
You might want to have a look into the stss box to see how many sync
samples there are.

Grüße,
Sebasztian


Chad Cabrera <notifications@github.com> schrieb am Mo., 24. Apr. 2017 um
04:02 Uhr:

**Comment by chad187** [comment]
I have tried simpleShortenedExample with double startTime1 = videoStart;
            double endTime1 = videoEnd; and shortenedExample as shown below.  Both will end up putting the start and end times at the same location.  I have used video length ranging from 10-40 seconds

`private void clipVideoLength(double videoStart, double videoEnd, String videoPath) {
        //shorten the video at the given location
        Movie movie = null;
        try {
            movie = MovieCreator.build(videoPath);

            List<Track> tracks = movie.getTracks();
            movie.setTracks(new LinkedList<Track>());
            // remove all tracks we will create new tracks from the old

            double startTime1 = videoStart;
            double endTime1 = videoEnd;
            double startTime2 = videoStart;
            double endTime2 = videoEnd;

            boolean timeCorrected = false;

            // Here we try to find a track that has sync samples. Since we can only start decoding
            // at such a sample we SHOULD make sure that the start of the new fragment is exactly
            // such a frame
            for (Track track : tracks) {
                if (track.getSyncSamples() != null && track.getSyncSamples().length > 0) {
                    if (timeCorrected) {
                        // This exception here could be a false positive in case we have multiple tracks
                        // with sync samples at exactly the same positions. E.g. a single movie containing
                        // multiple qualities of the same video (Microsoft Smooth Streaming file)

                        throw new RuntimeException("The startTime has already been corrected by another track with SyncSample. Not Supported.");
                    }
                    startTime1 = correctTimeToSyncSample(track, startTime1, false);
                    endTime1 = correctTimeToSyncSample(track, endTime1, true);
                    startTime2 = correctTimeToSyncSample(track, startTime2, false);
                    endTime2 = correctTimeToSyncSample(track, endTime2, true);
                    timeCorrected = true;
                }
            }

            for (Track track : tracks) {
                long currentSample = 0;
                double currentTime = 0;
                double lastTime = -1;
                long startSample1 = -1;
                long endSample1 = -1;
                long startSample2 = -1;
                long endSample2 = -1;

                for (int i = 0; i < track.getSampleDurations().length; i++) {
                    long delta = track.getSampleDurations()[i];


                    if (currentTime > lastTime && currentTime <= startTime1) {
                        // current sample is still before the new starttime
                        startSample1 = currentSample;
                    }
                    if (currentTime > lastTime && currentTime <= endTime1) {
                        // current sample is after the new start time and still before the new endtime
                        endSample1 = currentSample;
                    }
                    if (currentTime > lastTime && currentTime <= startTime2) {
                        // current sample is still before the new starttime
                        startSample2 = currentSample;
                    }
                    if (currentTime > lastTime && currentTime <= endTime2) {
                        // current sample is after the new start time and still before the new endtime
                        endSample2 = currentSample;
                    }
                    lastTime = currentTime;
                    currentTime += (double) delta / (double) track.getTrackMetaData().getTimescale();
                    currentSample++;
                }
                movie.addTrack(new AppendTrack(new CroppedTrack(track, startSample1, endSample1), new CroppedTrack(track, startSample2, endSample2)));
            }
            long start1 = System.currentTimeMillis();
            Container out = new DefaultMp4Builder().build(movie);
            long start2 = System.currentTimeMillis();
            FileOutputStream fos = new FileOutputStream(String.format("output-%f-%f--%f-%f.mp4", startTime1, endTime1, startTime2, endTime2));
            FileChannel fc = fos.getChannel();
            out.writeContainer(fc);

            fc.close();
            fos.close();
            long start3 = System.currentTimeMillis();
            System.err.println("Building IsoFile took : " + (start2 - start1) + "ms");
            System.err.println("Writing IsoFile took  : " + (start3 - start2) + "ms");
            System.err.println("Writing IsoFile speed : " + (new File(String.format("output-%f-%f--%f-%f.mp4", startTime1, endTime1, startTime2, endTime2)).length() / (start3 - start2) / 1000) + "MB/s");

        }catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }`

`private void joinVideos(){
       //join previous and next videos
        try {
            String[] videoUris = new String[]{//these files can't be right

                    mPreviousVideoAbsolutePath,
                    mNextVideoAbsolutePath

            };

            List<Movie> inMovies = new ArrayList<Movie>();
            for (String videoUri : videoUris) {
                inMovies.add(MovieCreator.build(videoUri));
            }

            List<Track> videoTracks = new LinkedList<Track>();
            List<Track> audioTracks = new LinkedList<Track>();

            for (Movie m : inMovies) {
                for (Track t : m.getTracks()) {
                    if (t.getHandler().equals("soun")) {
                        audioTracks.add(t);
                    }
                    if (t.getHandler().equals("vide")) {
                        videoTracks.add(t);
                    }
                }
            }

            Movie result = new Movie();

            if (audioTracks.size() > 0) {
                result.addTrack(new AppendTrack(audioTracks.toArray(new Track[audioTracks.size()])));
            }
            if (videoTracks.size() > 0) {
                result.addTrack(new AppendTrack(videoTracks.toArray(new Track[videoTracks.size()])));
            }

            Container out = new DefaultMp4Builder().build(result);

            FileChannel fc = new RandomAccessFile(String.format("output.mp4"), "rw").getChannel();
            out.writeContainer(fc);
            fc.close();
        }catch (IOException e){
            e.printStackTrace();
        }
    }`

`private static double correctTimeToSyncSample(Track track, double cutHere, boolean next) {
        double[] timeOfSyncSamples = new double[track.getSyncSamples().length];
        long currentSample = 0;
        double currentTime = 0;
        for (int i = 0; i < track.getSampleDurations().length; i++) {
            long delta = track.getSampleDurations()[i];

            if (Arrays.binarySearch(track.getSyncSamples(), currentSample + 1) >= 0) {
                // samples always start with 1 but we start with zero therefore +1
                timeOfSyncSamples[Arrays.binarySearch(track.getSyncSamples(), currentSample + 1)] = currentTime;
            }
            currentTime += (double) delta / (double) track.getTrackMetaData().getTimescale();
            currentSample++;

        }
        double previous = 0;
        for (double timeOfSyncSample : timeOfSyncSamples) {
            if (timeOfSyncSample > cutHere) {
                if (next) {
                    return timeOfSyncSample;
                } else {
                    return previous;
                }
            }
            previous = timeOfSyncSample;
        }
        return timeOfSyncSamples[timeOfSyncSamples.length - 1];
    }`
